File: lesson2/wangfei/lianxi/zoutaijie.go
package main

import (
	"fmt"
	"flag"
	"time"
)

func fib1(n int) int {
	if n == 1 {
		return 1
	}

	if n == 2 {
		return 2
	}

	return fib1(n-1) + fib1(n-2)
}

func fib2(n int) int {
	var (
		x     int
		y     int
		count int
	)

	if n == 1 {
		return 1
	}

	if n == 2 {
		return 2
	}

	x = 1
	y = 2
	for i := 2; i < n; i++ {
		count = x + y
		x, y = y, count
	}
	return count
}

func main() {
	var num int
	flag.IntVar(&num, "n", 0, "the num of fib")
	flag.Parse()

	start1 := time.Now()
	var res = fib1(num)
	t1 := time.Now()
	elapsed1 := t1.Sub(start1)
	fmt.Printf("递归算法,num:%d,计算结果:%d,耗时:%v\n", num,res,elapsed1)

	start2 := time.Now()
	var res2 = fib2(num)
	t2 := time.Now()
	elapsed2 := t2.Sub(start2)
	fmt.Printf("循环算法,num:%d,计算结果:%d,耗时:%v\n",num, res2,elapsed2)
}

File: lesson2/wangfei/lianxi/zoutaijie.py
#!-*- coding:utf-8 -*-
#!/usr/bin/python

import datetime
import sys


def fib1(n):
    if n == 1:
        return 1
    if n == 2:
        return 2

    return fib1(n-1)+fib1(n-2)

def fib2(n):
    if n == 1:
        return 1
    if n == 2:
        return 2
    if n>2:
        x,y=1,2
        for line in range(3,n+1):
            count = x + y
            x, y = y, count
    return count

def main():
    num = int(sys.argv[1])
	
    begin1 = datetime.datetime.now()
    res1 = fib1(num)
    end1 = datetime.datetime.now()
    elapsed1 = end1 - begin1

    begin2 = datetime.datetime.now()
    res2 = fib2(num)
    end2 = datetime.datetime.now()
    elapsed2 = end2 - begin2

    print "递归算法,num:{num},计算结果:{res},耗时:{elapsed1}ms".format(num=num,res=res1,elapsed1=elapsed1.microseconds)
    print "循环算法,num:{num},计算结果:{res},耗时:{elapsed2}ms".format(num=num,res=res2,elapsed2=elapsed2.microseconds)

if __name__ == "__main__":
    main()
